modelgym/gp_trainer.py

```diff
@@ -0,0 +1,122 @@
+        self.gp_evals, self.gp_eval_num = gp_evals, 0
+        self.default_params, self.best_params = None, None
+        self.best_n_estimators = None
+
+    def print_result(self, result, name='', extra_keys=None):
+        # TODO test
+        print('%s:\n' % name)
+        print('loss = %s' % (result['loss']))
+        if 'best_n_estimators' in result.keys():
+            print('best_n_estimators = %s' % result['best_n_estimators'])
+        elif 'n_estimators' in result.keys():
+            print('n_estimators = %s' % result['n_estimators'])
+        print('params = %s' % result['params'])
+        if extra_keys is not None:
+            for k in extra_keys:
+                if k in result:
+                    print("%s = %f" % (k, result[k]))
+
+    def crossval_optimize_params(self, model, cv_pairs, max_evals=None, verbose=True, batch_size=10,
+                                 tracker=None):
+        random_state = np.random.RandomState(1)
+        max_evals = max_evals or self.gp_evals
+        self.gp_eval_num, self.best_loss = 0, np.inf
+
+        prgp = modelgym.util.process_params_gp(model)
+        print("PRGP", prgp)
+
+        _ = gp_minimize(
+            func=lambda params: self.crossval_fit_eval(model=model, cv_pairs=cv_pairs, params=params, verbose=verbose),
+            dimensions=(prgp.values()), random_state=random_state, n_calls=max_evals,
+            n_jobs=max_evals - 1)
+
+        best_hyper_params = {k: v for k, v in zip(prgp.keys(), _.x)}
+        print(best_hyper_params)
+        bst = 1 - _.fun
+        print("Best accuracy score =", bst)
+        ans = best_hyper_params.copy()
+        ans['loss'] = bst
+        return ans if not isinstance(ans, SON) else ans.to_dict()
+
+    def crossval_fit_eval(self, model, cv_pairs, params=None, n_estimators=None, verbose=False):
+        params = params or model.default_params
+        if (isinstance(params, list)):
+            eta, max_depth, subsample, colsample_bytree, colsample_bylevel, min_child_weight, gamma, alpha, lambdax = \
+                params[0], params[1], params[2], params[3], params[4], params[5], params[6], params[7], params[8]
+            model.set_params(eta=eta,
+                             max_depth=max_depth,
+                             subsample=subsample,
+                             colsample_bylevel=colsample_bylevel,
+                             colsample_bytree=colsample_bytree,
+                             min_child_weight=min_child_weight,
+                             gamma=gamma,
+                             alpha=alpha,
+                             lambdax=lambdax
+                             )
+        n_estimators = n_estimators or self.n_estimators
+        evals_results, start_time = [], time.time()
+        if (isinstance(params, dict)):
+            params = model.preprocess_params(params)
+
+        mean_evals_results = []
+        std_evals_results = []
+
+        for dtrain, dtest in cv_pairs:
```

Comment:
looks like pretty generic piece of code. can you re-use base method of Trainer either here or at `fit`?